### PR TITLE
Added ViaRequest interceptor

### PR DESCRIPTION
**Pull request by arturobernalg**

### CODE REVIEW

**Comment by ok2c** [comment]
@arturobernalg Please change the target branch to `5.3.x`.

**Comment by arturobernalg** [comment]
> @arturobernalg Please change the target branch to `5.3.x`.

@ok2c Done

**Review by ok2c — CHANGES_REQUESTED** [review]

File: httpcore5/src/main/java/org/apache/hc/core5/http/protocol/ViaRequest.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+     * @throws ProtocolException if there was a protocol error
+     * @throws IOException if there was an I/O error
+     */
+    @Override
+    public void process(final HttpRequest request, final EntityDetails entity, final HttpContext context) throws ProtocolException, IOException {
+        Args.notNull(request, "HTTP request");
+        Args.notNull(context, "HTTP context");
+        final ProtocolVersion ver = context.getProtocolVersion() != null ? context.getProtocolVersion() : HttpVersion.HTTP_1_1;
+
+        final Header host = request.getHeader(HttpHeaders.HOST);
```
[the remark this answers is not in the document]
Comment (arturobernalg) [reply httpcore5/src/main/java/org/apache/hc/core5/http/protocol/ViaRequest.java]: Sounds reasonable.. What about now @ok2c ?

File: httpcore5/src/main/java/org/apache/hc/core5/http/protocol/ViaRequest.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+       if (!ver.greaterEquals(HttpVersion.HTTP_1_1)) {
+           throw new ProtocolException("Invalid protocol version: %s", ver);
+       }
+       String viaHeaderValue = ver.getProtocol() + " " + ver.getMajor() + "." + ver.getMinor() + " " + authority.getHostName();
+       final int port = authority.getPort();
+       if (port != -1) {
+            viaHeaderValue += ":" + port;
+        }
+        if (!request.containsHeader(HttpHeaders.VIA)) {
```
Comment (ok2c) [inline httpcore5/src/main/java/org/apache/hc/core5/http/protocol/ViaRequest.java:122]: @arturobernalg One last thing. Please move the `Via` header value generation into this if statement. There is not point going through the whole process just to find out the value is not needed after all.

Comment (arturobernalg) [reply httpcore5/src/main/java/org/apache/hc/core5/http/protocol/ViaRequest.java:122]: @ok2c 
Something like this
` if (!request.containsHeader(HttpHeaders.VIA)) {
            String viaHeaderValue = ver.getProtocol() + " " + ver.getMajor() + "." + ver.getMinor() + " " + authority.getHostName();
            final int port = authority.getPort();
            if (port != -1) {
                viaHeaderValue += ":" + port;
            }
            request.addHeader(HttpHeaders.VIA, viaHeaderValue);
        }`

Comment (ok2c) [reply httpcore5/src/main/java/org/apache/hc/core5/http/protocol/ViaRequest.java:122]: @arturobernalg Something like that.

File: httpcore5/src/main/java/org/apache/hc/core5/http/protocol/ViaRequest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ok2c) [reply httpcore5/src/main/java/org/apache/hc/core5/http/protocol/ViaRequest.java]: Merged. Thank you, @arturobernalg, for this contribution.

**Pull request merged**